catalogue_graph/src/reporter.py

```diff
@@ -0,0 +1,192 @@
+
+    # open with smart_open, check for file existence
+    try:
+        with smart_open.open(trigger_report_url, "r") as f:
+            trigger_report = TriggerReport.model_validate_json(f.read())
+        with smart_open.open(loader_report_url, "r") as f:
+            loader_report = LoaderReport.model_validate_json(f.read())
+        with smart_open.open(previous_loader_report_url, "r") as f:
+            previous_loader_report = LoaderReport.model_validate_json(f.read())
+    # if file does not exist, ignore
+    except (OSError, KeyError) as e:
+        print(f"Report not found: {e}")
+
+    final_report = FinalReport(
+        pipeline_date=pipeline_date,
+        job_id=job_id,
+        neptune_record_count=trigger_report.record_count,
+        es_record_count=loader_report.record_count,
+        success_count=event.success_count,
+        previous_es_record_count=previous_loader_report.record_count,
+        previous_job_id=previous_loader_report.job_id,
+    )
+
+    start_datetime = dateTimeFromJobId(job_id)
+    previous_start_datetime = dateTimeFromJobId(final_report.previous_job_id)
+    time_delta = round(
+        divmod(
+            (datetime.now() - datetime.strptime(job_id, "%Y%m%dT%H%M")).total_seconds(),
+            60,
+        )[0]
+    )
+
+    if final_report.neptune_record_count == final_report.success_count:
+        graph_index_comparison = "_(the same as the graph)_"
+    else:
+        graph_index_comparison = (
+            f":warning: _compared to {final_report.neptune_record_count} in the graph_"
+        )
+
+    return [
+        {
+            "type": "header",
+            "text": {
+                "type": "plain_text",
+                "emoji": True,
+                "text": f"{':white_check_mark:'} Concepts :bulb:",
+            },
+        },
+        {
+            "type": "section",
+            "text": {
+                "type": "mrkdwn",
+                "text": "\n".join(
+                    [
+                        f"- Index *concepts-indexed-{pipeline_date}*",
+                        f"- Pipeline started *{start_datetime}*",
+                        f"- It contains *{final_report.success_count}* documents {graph_index_comparison}.",
+                        f"- Pipeline took *{time_delta} minutes* to complete.",
+                        f"- The last update was {previous_start_datetime}when {final_report.previous_es_record_count} documents were indexed.",
+                    ]
+                ),
+            },
+        },
+    ]
+
+
+def publish_report(report: list[typing.Any], config: ReporterConfig) -> None:
+    secretsmanager = boto3.Session().client("secretsmanager")
+    slack_endpoint = secretsmanager.get_secret_value(SecretId=config.slack_secret)[
+        "SecretString"
+    ]
+
+    httpx.post(slack_endpoint, json={"blocks": report})
+
+
+def handler(event: ReporterEvent, context: typing.Any) -> None:
```

Comment:
I'd repeat pattern in the other ingestor lambdas where there is a `lambda_handler` function, that performs `model_validate` on the incoming event (which will also fill in the default values if required), and allows you to control the config handed to the handler function in testing.